firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java

```diff
@@ -193,8 +194,26 @@ WriteStream createWriteStream(WriteStream.Callback listener) {
               for (DocumentKey key : keys) {
                 results.add(resultMap.get(key));
               }
-              return results;
-            });
+              tcs.setResult(results);
+            }
+          }
+
+          @Override
+          public void onClose(Status status) {
+            if (status.isOk() && !callback_fired) {
+              callback_fired = true;
+              tcs.setResult(new ArrayList<>());
```

Comment:
`Collections.emptyList()`, which just returns a reference to an already allocated object.

Reply:
FWIW you can also remove `callback_fired` and invoke `tcs.trySetResult()`.